### PR TITLE
VIH-6002 Get a list of distinct judge first names

### PR DESCRIPTION
**Pull request by jdssunil**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/VIH-6002

### Change description ###
Get a list of distinct judge first names

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ x] No
```

### CODE REVIEW

File: VideoAPI/Video.API/Controllers/ParticipantsController.cs [hunk 1]
```diff
@@ -266,5 +266,20 @@ public async Task<IActionResult> SaveHeartbeatDataForParticipantAsync(Guid confe
 
             return NoContent();
         }
+
+        /// <summary>
+        /// Get a list of distinct first name of judges
+        /// </summary>
+        /// <returns></returns>
+        [HttpGet("participants/judge/firstname")]
```
Comment (shaed-parkar) [inline VideoAPI/Video.API/Controllers/ParticipantsController.cs:274]: judge should be plural since it's a collection of resources

Comment (jdssunil) [reply VideoAPI/Video.API/Controllers/ParticipantsController.cs]: Fixed

File: VideoAPI/VideoApi.AcceptanceTests/Steps/ConferenceSteps.cs [hunk 2]
```diff
@@ -251,6 +265,32 @@ public void ThenTheJudgeInHearingResponseShouldBeRetrieved()
             expectedHearing.Username.Should().Be(judge.Username);
         }
 
+        private void CreateMultipleConferences(bool addDuplicateFirstNames)
+        {
+            _context.Test.ConferenceDetailsResponses = new List<ConferenceDetailsResponse>();
+            string judge1 = "Automation_" + Name.First() + RandomNumber.Next();
+            string judge2 = "Automation_" + Name.First() + RandomNumber.Next();
+            
+            for (int i = 0; i < 2; i++)
```
Comment (npoxon) [inline VideoAPI/VideoApi.AcceptanceTests/Steps/ConferenceSteps.cs:274]: you have hard coded this so it will only ever make 2 conferences. Can you reuse the existing method for creating a single hearing multiple times and pass in the name you want to duplicate? Better that than duplicating the code.

File: VideoAPI/Testing.Common/Helper/Builders/Api/BookNewConferenceRequestBuilder.cs [hunk 3]
```diff
@@ -25,17 +25,23 @@ public BookNewConferenceRequestBuilder(string caseName)
                 .Build();
         }
 
-        public BookNewConferenceRequestBuilder WithJudge()
+        public BookNewConferenceRequestBuilder WithJudge(string firstName = null)
         {
             var participant = Builder<ParticipantRequest>.CreateNew()
-                .With(x => x.Name = $"Automation_{Name.FullName()}")
+                .With(x => x.Name = "Automation_" + Name.First() + Faker.RandomNumber.Next())
                 .With(x => x.FirstName = $"Automation_{Name.First()}")
```
Comment (npoxon) [inline VideoAPI/Testing.Common/Helper/Builders/Api/BookNewConferenceRequestBuilder.cs:32]: can you match all the other sting formats? $"Automation_{Name.First()}{Faker.RandomNumber.Next()}"

File: VideoAPI/VideoApi.AcceptanceTests/Features/Participants.feature [hunk 4]
```diff
@@ -39,3 +39,19 @@ Feature: Participants
...
+    Then the response should have the status Ok and success status True
+    And the judge names should be retrieved
+
+  @VIH-6002
+  Scenario: Get the a distinct list of judge names
+    Given I have multiple conferences with duplicate first names for judges
+    And I have a valid get judge names data request 
+    When I send the request to the endpoint
+    Then the response should have the status Ok and success status True
+    And the judge names should be retrieved
```
Comment (npoxon) [inline VideoAPI/VideoApi.AcceptanceTests/Features/Participants.feature:57]: Whats the difference between these 2 tests? Can we not just have the 2nd one?

File: VideoAPI/VideoApi.AcceptanceTests/Steps/ParticipantsSteps.cs [hunk 5]
```diff
@@ -141,5 +141,28 @@ public void ThenTheParticipantIsAdded(string state)
...
+            var judgesList = RequestHelper.DeserialiseSnakeCaseJsonToResponse<JudgeNameListResponse>(_context.Response.Content).FirstNames;
+            var conferences = _context.Test.ConferenceDetailsResponses;
+
+            var judges = conferences.SelectMany(c => c.Participants).Where(p => p.UserRole == UserRole.Judge);
+
+            foreach (var judge in judges)
+            {
+                judgesList.Count(x => x.Contains(judge.FirstName)).Should().Be(1);
+                judgesList.Count.Should().BeGreaterOrEqualTo(2);
+
```
Comment (npoxon) [inline VideoAPI/VideoApi.AcceptanceTests/Steps/ParticipantsSteps.cs:163]: but you are not validating you are getting the _right_ judges?

File: VideoAPI/VideoApi.IntegrationTests/Database/Queries/GetDistinctJudgeListByFirstNameQueryTests.cs [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+            var judgelist = await _handler.Handle(new GetDistinctJudgeListByFirstNameQuery());
+
+            judgelist.Should().NotBeEmpty();
+        }
+
+        [Test]
+        public async Task Should_get_unique_list_of_judges()
+        {
+            //This is done to create an unique first name to avoid the test from failing because of any existing data
+            string judge1 = "Automation_" + Name.First() + Faker.RandomNumber.Next();
```
Comment (npoxon) [inline VideoAPI/VideoApi.IntegrationTests/Database/Queries/GetDistinctJudgeListByFirstNameQueryTests.cs:53]: Can you remove the comment and instead make it a function? CreateUniqueName()

File: VideoAPI/VideoApi.UnitTests/Controllers/Participant/GetDistinctJudgeNamesTests.cs [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        {
+            var judgeFirstnames = new List<string> { "JudgeName1", "JudgeName2", "JudgeName3" };
+
+            MockQueryHandler
+                .Setup(x => x.Handle<GetDistinctJudgeListByFirstNameQuery, List<string>>(It.IsAny<GetDistinctJudgeListByFirstNameQuery>()))
+                .ReturnsAsync(judgeFirstnames);
+        }
+
+        [Test]
+        public async Task should_getjudgelistResponses()
```
Comment (npoxon) [inline VideoAPI/VideoApi.UnitTests/Controllers/Participant/GetDistinctJudgeNamesTests.cs:27]: are you using the linting? Methods should start with a capital letter and you have a mix of camelcase and underscores

Comment (npoxon) [reply VideoAPI/VideoApi.UnitTests/Controllers/Participant/GetDistinctJudgeNamesTests.cs:27]: Should_get_judge_list_response or ShouldGetJudgeListResponse

File: VideoAPI/VideoApi.AcceptanceTests/Steps/ParticipantsSteps.cs (hunk 5 above)
Comment (npoxon) [reply VideoAPI/VideoApi.AcceptanceTests/Steps/ParticipantsSteps.cs:163]: Also there is an empty line

File: VideoAPI/VideoApi.UnitTests/Controllers/Participant/GetDistinctJudgeNamesTests.cs [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+            result.Should().NotBeNull();
+            result.Should().BeAssignableTo<OkObjectResult>();
+            var response = result.As<OkObjectResult>().Value.As<JudgeNameListResponse>();
+            response.FirstNames.Count.Should().Be(3);
+        }
+    }
+
+}
+
```
Comment (npoxon) [inline VideoAPI/VideoApi.UnitTests/Controllers/Participant/GetDistinctJudgeNamesTests.cs:41]: Can you add test for the negative scenarios as well?

File: VideoAPI/VideoApi.AcceptanceTests/Steps/ConferenceSteps.cs (hunk 2 above)
Comment (jdssunil) [reply VideoAPI/VideoApi.AcceptanceTests/Steps/ConferenceSteps.cs:274]: Well spotted

File: VideoAPI/VideoApi.AcceptanceTests/Features/Participants.feature (hunk 4 above)
Comment (jdssunil) [reply VideoAPI/VideoApi.AcceptanceTests/Features/Participants.feature:57]: Removed the first one

File: VideoAPI/Testing.Common/Helper/Builders/Api/BookNewConferenceRequestBuilder.cs (hunk 3 above)
Comment (jdssunil) [reply VideoAPI/Testing.Common/Helper/Builders/Api/BookNewConferenceRequestBuilder.cs:32]: Done

File: VideoAPI/VideoApi.IntegrationTests/Database/Queries/GetDistinctJudgeListByFirstNameQueryTests.cs (hunk 6 above)
Comment (jdssunil) [reply VideoAPI/VideoApi.IntegrationTests/Database/Queries/GetDistinctJudgeListByFirstNameQueryTests.cs:53]: Good idea :)
